File: apps/api/engine/cyclonedx.py
"""
CycloneDX 1.6 AI-BOM (Artificial Intelligence Bill of Materials) output formatter.
Produces an SBOM-compatible document documenting MCP server components.
"""
from datetime import datetime, timezone
from .models import ScanResult, Severity
from .parser import MCPConfig


def to_cyclonedx(result: ScanResult, config: MCPConfig) -> dict:
    now = datetime.now(timezone.utc).isoformat()

    components: list[dict] = []
    for server in config.servers:
        package = _extract_package(server)
        comp: dict = {
            "type": "library",
            "bom-ref": f"mcp-server:{server.name}",
            "name": server.name,
            "description": f"MCP server — command: {server.command or 'n/a'}",
            "properties": [
                {"name": "mcp:transport", "value": server.transport or "stdio"},
                {"name": "mcp:command", "value": server.command or ""},
            ],
        }

        if package:
            pkg_name, pkg_version = _split_package(package)
            comp["purl"] = _build_purl(server, pkg_name, pkg_version)
            if pkg_name:
                comp["name"] = pkg_name
            if pkg_version:
                comp["version"] = pkg_version
            else:
                comp["version"] = "unpinned"
                comp["properties"].append(
                    {"name": "mcp:version-pinned", "value": "false"}
                )

        if server.url:
            comp["properties"].append({"name": "mcp:url", "value": server.url})

        # Map findings to vulnerabilities section
        server_findings = [f for f in result.findings if f.server_name == server.name]
        if server_findings:
            comp["properties"].append(
                {"name": "mcp:finding-count", "value": str(len(server_findings))}
            )

        components.append(comp)

    # Collect vulnerabilities from scan result
    vulnerabilities: list[dict] = []
    for finding in result.findings:
        vuln: dict = {
            "bom-ref": f"vuln:{finding.check_id}:{finding.server_name}",
            "id": finding.check_id,
            "source": {
                "name": "MCPAudit",
                "url": f"https://mcpaudit.app/checks/{finding.check_id}",
            },
            "ratings": [
                {
                    "source": {"name": "MCPAudit"},
                    "severity": finding.severity.value,
                }
            ],
            "description": finding.detail,
            "recommendation": finding.remediation,
            "properties": [
                {"name": "owasp-mcp", "value": finding.owasp.value},
                {"name": "mcp:server", "value": finding.server_name},
                {"name": "mcp:engine", "value": finding.engine},
            ],
            "affects": [
                {
                    "ref": f"mcp-server:{finding.server_name}",
                }
            ],
        }
        if finding.attack_tactic:
            vuln["properties"].append({"name": "attack:tactic", "value": finding.attack_tactic})
        vulnerabilities.append(vuln)

    return {
        "bomFormat": "CycloneDX",
        "specVersion": "1.6",
        "serialNumber": f"urn:uuid:{result.scan_id}",
        "version": 1,
        "metadata": {
            "timestamp": now,
            "tools": [
                {
                    "type": "application",
                    "vendor": "MCPAudit",
                    "name": "MCPAudit",
                    "version": "0.1.0",
                    "externalReferences": [
                        {"type": "website", "url": "https://mcpaudit.app"}
                    ],
                }
            ],
            "component": {
                "type": "application",
                "name": "MCP Configuration",
                "description": f"Scanned MCP config — hash {result.config_hash}",
                "properties": [
                    {"name": "mcp:servers-count", "value": str(result.summary.servers_scanned)},
                    {"name": "mcp:scan-id", "value": result.scan_id},
                ],
            },
        },
        "components": components,
        "vulnerabilities": vulnerabilities,
    }


def _extract_package(server) -> str | None:
    if server.command in ("npx", "npm", "uvx", "pip", "pip3"):
        for arg in server.args:
            if not arg.startswith("-"):
                return arg
    return None


def _split_package(package: str) -> tuple[str, str]:
    if package.startswith("@"):
        parts = package.split("@")
        if len(parts) >= 3:
            name = "@" + parts[1]
            if "/" in parts[1]:
                name = "@" + parts[1]
            version = parts[2] if len(parts) > 2 else ""
            return name, version
        return package, ""
    if "@" in package:
        idx = package.index("@")
        return package[:idx], package[idx + 1:]
    return package, ""


def _build_purl(server, pkg_name: str, pkg_version: str) -> str:
    if server.command in ("npx", "npm"):
        ecosystem = "npm"
    elif server.command in ("uvx", "pip", "pip3"):
        ecosystem = "pypi"
    else:
        return ""
    version_part = f"@{pkg_version}" if pkg_version else ""
    # URL-encode @ in scoped npm package names
    encoded_name = pkg_name.replace("@", "%40").replace("/", "%2F") if pkg_name.startswith("@") else pkg_name
    return f"pkg:{ecosystem}/{encoded_name}{version_part}"

File: packages/cli/internal/engine/cyclonedx.go
package engine

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/abhijitdalal26/MCP-Audit/cli/internal/engine/models"
)

type bomProperty struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type bomSource struct {
	Name string `json:"name"`
	URL  string `json:"url,omitempty"`
}

type bomRating struct {
	Source   bomSource `json:"source"`
	Severity string    `json:"severity"`
}

type bomAffect struct {
	Ref string `json:"ref"`
}

type bomVuln struct {
	BOMRef         string        `json:"bom-ref"`
	ID             string        `json:"id"`
	Source         bomSource     `json:"source"`
	Ratings        []bomRating   `json:"ratings"`
	Description    string        `json:"description"`
	Recommendation string        `json:"recommendation"`
	Properties     []bomProperty `json:"properties"`
	Affects        []bomAffect   `json:"affects"`
}

type bomComponent struct {
	Type        string        `json:"type"`
	BOMRef      string        `json:"bom-ref"`
	Name        string        `json:"name"`
	Description string        `json:"description"`
	PURL        string        `json:"purl,omitempty"`
	Version     string        `json:"version,omitempty"`
	Properties  []bomProperty `json:"properties"`
}

// marshalBOM produces a CycloneDX 1.6 AI-BOM JSON document.
// The function needs access to the parsed config to build the components list.
// We store it alongside the ScanResult via the helper below.
func marshalBOM(result *models.ScanResult) ([]byte, error) {
	// We do not have the config object here — BOM is generated from what ScanResult records.
	// Build minimal components from the findings' server names.
	seenServers := map[string]bool{}
	for _, f := range result.Findings {
		seenServers[f.ServerName] = true
	}

	components := []bomComponent{}
	for name := range seenServers {
		if name == "(all servers)" || strings.HasPrefix(name, "(") {
			continue
		}
		comp := bomComponent{
			Type:        "library",
			BOMRef:      "mcp-server:" + name,
			Name:        name,
			Description: fmt.Sprintf("MCP server — name: %s", name),
			Properties:  []bomProperty{},
		}
		count := 0
		for _, f := range result.Findings {
			if f.ServerName == name {
				count++
			}
		}
		if count > 0 {
			comp.Properties = append(comp.Properties, bomProperty{
				Name: "mcp:finding-count", Value: fmt.Sprintf("%d", count),
			})
		}
		components = append(components, comp)
	}

	vulns := []bomVuln{}
	for _, f := range result.Findings {
		props := []bomProperty{
			{Name: "owasp-mcp", Value: string(f.OWASP)},
			{Name: "mcp:server", Value: f.ServerName},
			{Name: "mcp:engine", Value: f.Engine},
		}
		if f.AttackTactic != "" {
			props = append(props, bomProperty{Name: "attack:tactic", Value: f.AttackTactic})
		}
		vulns = append(vulns, bomVuln{
			BOMRef:         fmt.Sprintf("vuln:%s:%s", f.CheckID, f.ServerName),
			ID:             f.CheckID,
			Source:         bomSource{Name: "MCPAudit", URL: fmt.Sprintf("https://mcpaudit.app/checks/%s", f.CheckID)},
			Ratings:        []bomRating{{Source: bomSource{Name: "MCPAudit"}, Severity: string(f.Severity)}},
			Description:    f.Detail,
			Recommendation: f.Remediation,
			Properties:     props,
			Affects:        []bomAffect{{Ref: "mcp-server:" + f.ServerName}},
		})
	}

	doc := map[string]any{
		"bomFormat":    "CycloneDX",
		"specVersion":  "1.6",
		"serialNumber": fmt.Sprintf("urn:uuid:%s", result.ScanID),
		"version":      1,
		"metadata": map[string]any{
			"timestamp": time.Now().UTC().Format(time.RFC3339),
			"tools": []map[string]any{{
				"type":   "application",
				"vendor": "MCPAudit",
				"name":   "MCPAudit",
				"version": "0.1.0",
				"externalReferences": []map[string]any{{
					"type": "website",
					"url":  "https://mcpaudit.app",
				}},
			}},
			"component": map[string]any{
				"type":        "application",
				"name":        "MCP Configuration",
				"description": fmt.Sprintf("Scanned MCP config — hash %s", result.ConfigHash),
				"properties": []bomProperty{
					{Name: "mcp:servers-count", Value: fmt.Sprintf("%d", result.Summary.ServersScanned)},
					{Name: "mcp:scan-id", Value: result.ScanID},
				},
			},
		},
		"components":      components,
		"vulnerabilities": vulns,
	}

	return json.MarshalIndent(doc, "", "  ")
}

